src/cluster/mcl.cpp

```diff
@@ -0,0 +1,570 @@
+				assert(m.size() <= order.size() * order.size());
+				float sparsity = 1.0-(1.0 * m.size()) / (order.size()*order.size());
+				max_sparsities[iThr] = max(max_sparsities[iThr], sparsity);
+				min_sparsities[iThr] = min(min_sparsities[iThr], sparsity);
+				// map back to original ids
+				vector<unordered_set<uint32_t>> list_of_sets;
+				unordered_set<uint32_t> attractors;
+
+				//TODO: a size limit for the dense matrix should control this as well
+#ifdef MCL_TIMINGS
+				chrono::high_resolution_clock::time_point t = chrono::high_resolution_clock::now();
+#endif
+				if(sparsity >= config.cluster_mcl_sparsity_switch && expansion - (int) expansion == 0){ 
+					n_sparse++;
+					Eigen::SparseMatrix<float> m_sparse(order.size(), order.size());
+					m_sparse.setFromTriplets(m.begin(), m.end());
+#ifdef MCL_TIMINGS
+					sparse_create_time += chrono::duration_cast<chrono::milliseconds>(chrono::high_resolution_clock::now() - t).count();
+#endif
+					markov_process(&m_sparse, inflation, expansion);
+#ifdef MCL_TIMINGS
+					chrono::high_resolution_clock::time_point t = chrono::high_resolution_clock::now();
+#endif
+					LazyDisjointIntegralSet<uint32_t> disjointSet(m_sparse.cols());
+					for (uint32_t k=0; k<m_sparse.outerSize(); ++k){
+						for (Eigen::SparseMatrix<float>::InnerIterator it(m_sparse, k); it; ++it){
+							assert(abs(it.value()) > numeric_limits<float>::epsilon());
+							disjointSet.merge(it.row(), it.col());
+							if(it.row() == it.col()){
+								attractors.emplace(it.row());
+							}
+						}
+					}
+					list_of_sets = disjointSet.getListOfSets();
+#ifdef MCL_TIMINGS
+					sparse_list_time += chrono::duration_cast<chrono::milliseconds>(chrono::high_resolution_clock::now() - t).count();
+#endif
+				}
+				else{
+					n_dense++;
+					Eigen::MatrixXf m_dense = Eigen::MatrixXf::Zero(order.size(), order.size());
+					for(Eigen::Triplet<float> const & t : m){
+						m_dense(t.row(), t.col()) = t.value();
+					}
+#ifdef MCL_TIMINGS
+					dense_create_time += chrono::duration_cast<chrono::milliseconds>(chrono::high_resolution_clock::now() - t).count();
+#endif
+					markov_process(&m_dense, inflation, expansion);
+#ifdef MCL_TIMINGS
+					chrono::high_resolution_clock::time_point t = chrono::high_resolution_clock::now();
+#endif
+					LazyDisjointIntegralSet<uint32_t> disjointSet(m_dense.cols());
+					for (uint32_t icol=0; icol<m_dense.cols(); ++icol){
+						for (uint32_t irow=0; irow<m_dense.rows(); ++irow){
+							if( abs(m_dense(irow, icol)) > numeric_limits<float>::epsilon()){
+								disjointSet.merge(irow, icol);
+								if(irow == icol){
+									attractors.emplace(irow);
+								}
+							}
+						}
+					}
+					list_of_sets = disjointSet.getListOfSets();
+#ifdef MCL_TIMINGS
+					dense_list_time += chrono::duration_cast<chrono::milliseconds>(chrono::high_resolution_clock::now() - t).count();
+#endif
+				}
+				for(unordered_set<uint32_t> subset : list_of_sets){
+					assert(cluster_id < 0x00ffffff);
```

Comment:
Do we support a max. of 16.7 million clusters?

Reply:
right now yes. This can be changed easily. One question is whether we want to be able to lable three or only two kinds of nodes, i.e. singletons (which may be inferred from the clusters), normal nodes and attractor nodes. Removing the singleton label gives a factor two, i.e. roughly 33 milion clusters, if that is not enough, we can just switch to a 64 bit integer instead (or bigger if necessary) at the cost of an increased memory footprint

Reply:
Maybe use a struct or a second array for this? Seems a bit cleaner.